### PR TITLE
[TASK] Return `ResponseInterface` in controller actions

### PR DESCRIPTION
**Pull request by linawolf**
resolves https://github.com/TYPO3-Documentation/tea/issues/564

### CODE REVIEW

File: Classes/Controller/TeaController.php [hunk 1]
```diff
@@ -18,18 +19,21 @@ class TeaController extends ActionController
      */
     private $teaRepository;
 
-    public function injectTeaRepository(TeaRepository $teaRepository): void
+    public function injectTeaRepository(TeaRepository $teaRepository): ResponseInterface
```
Comment (brotkrueml) [inline Classes/Controller/TeaController.php:22]: The inject method does not return anything, it is for injecting a dependency. Is there a reason not using constructor injection?

Comment (DavidBruchmann) [reply Classes/Controller/TeaController.php:22]: the inject method never has to return a ResponseInterface I suppose.
can be the old value void
[3:05](https://typo3.slack.com/archives/CSQ8T0VS5/p1664395534686369)
only actions return a ResponseInterface afaik

Comment (sbuerk) [reply Classes/Controller/TeaController.php:22]: inject methods should be return type void and not returning anything. Does not make any sense anyway.

File: Tests/Unit/Controller/TeaControllerTest.php [hunk 2]
```diff
@@ -51,6 +58,11 @@ protected function setUp(): void
         /** @var TeaRepository&ProphecySubjectInterface $teaRepository */
         $teaRepository = $this->teaRepositoryProphecy->reveal();
         $this->subject->injectTeaRepository($teaRepository);
+
+        $this->responseFactoryProphecy = $this->prophesize(ResponseFactory::class);
+        /** @var ResponseFactory&ProphecySubjectInterface $responseFactory */
+        $responseFactory = $this->responseFactoryProphecy->reveal();
```
Comment (sbuerk) [inline Tests/Unit/Controller/TeaControllerTest.php:64]: `$this->subject->injectTeaRepositoy()` is changed to return `$this->htmlResponse()` .. which is wrong anyway. You added code to inject the responseFactory .. but "to late". The responseFactory must be existing in the actionController before first usage (which is added in `injectTeaRepository` ...

File: Tests/Unit/Controller/TeaControllerTest.php [hunk 3]
```diff
@@ -69,6 +81,7 @@ public function indexActionAssignsAllTeaAsTeasToView(): void
         $teas = $this->prophesize(QueryResultInterface::class)->reveal();
         $this->teaRepositoryProphecy->findAll()->willReturn($teas);
         $this->viewProphecy->assign('teas', $teas)->shouldBeCalled();
+        $this->responseFactoryProphecy->createResponse()->shouldBeCalled();
```
Comment (sbuerk) [inline Tests/Unit/Controller/TeaControllerTest.php:84]: Eventually add here a `->willReturn(new HtmlResponse())` or a proper prophesized response or setup response. Eventually do not prophecies the response factory if possibile (not looked into it)

File: Classes/Controller/TeaController.php (hunk 1 above)
Comment (linawolf) [reply Classes/Controller/TeaController.php]: yeah, I was kind of blind

File: Tests/Unit/Controller/TeaControllerTest.php (hunk 3 above)
Comment (linawolf) [reply Tests/Unit/Controller/TeaControllerTest.php]: I didn't find a way not to prophesize this and still get the tests running However it feels like I am testing Core Details now and not really the Tea Controller...

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,7 @@ This project adheres to [Semantic Versioning](https://semver.org/).
 
 ### Changed
 - Replace switchable controller actions with separate plugins (#575)
+- Controller actions return ResponseInterface (#597)
```
Comment (oliverklee) [inline CHANGELOG.md:12]: Please move the entry to the top of the section so the changes are in reverse chronological order ("newest on top").

Please use the same format in the changelog as in the commit message subject, i.e., use the imperative:

Return `ResponseInterface` in controller actions

For details on improving git commit message, I can highly recommend this article: https://cbea.ms/git-commit/

File: Tests/Unit/Controller/TeaControllerTest.php [hunk 5]
```diff
@@ -51,6 +65,16 @@ protected function setUp(): void
...
+
+        $this->responseFactoryProphecy = $this->prophesize(ResponseFactory::class);
+        /** @var ResponseFactory&ProphecySubjectInterface $responseFactory */
+        $responseFactory = $this->responseFactoryProphecy->reveal();
+        $this->subject->injectResponseFactory($responseFactory);
+
+        $this->streamFactoryProphecy = $this->prophesize(StreamFactory::class);
+        /** @var StreamFactory&ProphecySubjectInterface $streamFactory */
+        $streamFactory = $this->streamFactoryProphecy->reveal();
+        $this->subject->injectStreamFactory($streamFactory);
```
Comment (oliverklee) [inline Tests/Unit/Controller/TeaControllerTest.php:77]: You can simplify this a lot by adding `htmlResponse` to the list of mocked methods in line 58.

File: Classes/Controller/TeaController.php [hunk 6]
```diff
@@ -23,13 +24,15 @@ public function injectTeaRepository(TeaRepository $teaRepository): void
         $this->teaRepository = $teaRepository;
     }
 
-    public function indexAction(): void
+    public function indexAction(): ResponseInterface
```
Comment (oliverklee) [inline Classes/Controller/TeaController.php:27]: We should also have tests that the actions return a `HtmlResponse` with the content from `htmlResponse`.

**Review by oliverklee — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 7]
```diff
@@ -8,8 +8,8 @@ This project adheres to [Semantic Versioning](https://semver.org/).
 ### Added
 
 ### Changed
+- Return ResponseInterface in controller actions (#597)
```
Comment (oliverklee) [inline CHANGELOG.md:11]: And to make the changelog even more beautiful, please put code, class names etc. (in this case, `ResponseInterface`) in backticks so it gets rendered in a monospace font. (GitHub parses Markdown in the release notes, too.)

File: Tests/Unit/Controller/TeaControllerTest.php (hunk 5 above)
Comment (linawolf) [reply Tests/Unit/Controller/TeaControllerTest.php]: Done

File: Classes/Controller/TeaController.php (hunk 6 above)
Comment (linawolf) [reply Classes/Controller/TeaController.php:27]: Done

**Review by oliverklee — APPROVED** [review]
LGTM.

**Pull request merged**